modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/LimitOffsetIntegrationTest.java

```diff
@@ -88,6 +88,16 @@ public class LimitOffsetIntegrationTest extends AbstractBasicIntegrationTest {
                     .setSqlSchema("PUBLIC"));
     }
 
+    /** */
+    @Test
+    public void testNestedLimitOffsetWithUnion() {
+        sql("INSERT into TEST_REPL VALUES (1, 'a'), (2, 'b'), (3, 'c'), (4, 'd')");
+
+        assertQuery("(SELECT id FROM TEST_REPL WHERE id = 2) UNION ALL " +
```

Comment:
Looks like UNION_ALL is not related to the problem and test can be simplified?

Reply:
It is. It changes `bb.top` condition and the problem not appears.